### PR TITLE
Only serialize blazor hub protocol for blazor hub

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes https://github.com/Azure/azure-signalr/issues/693

I think master is the right branch for 1.1.x?

### CODE REVIEW

**Comment by BrennanConroy** [comment]
@vicancy @halter73 @anurse @davidfowl

File: src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.SignalR.Protocol;
+using Microsoft.AspNetCore.SignalR;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class DefaultHubMessageSerializer
```
Comment (anurse) [inline src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs]: Is this the code you copied from SignalR server? Can you put a comment in saying which commit it's from, etc. (like, "Copied from [full GitHub URL]" or something). I'd like to be able to cross-reference this later.

In 5.0 we plan to look at introducing a more public "API" for this so we don't have to copy code around :).

File: src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                }
+            }
+        }
+
+        public IReadOnlyList<SerializedMessage> SerializeMessage(HubMessage message)
+        {
+            var list = new List<SerializedMessage>(_hubProtocols.Count);
+            foreach (var protocol in _hubProtocols)
+            {
+                if (_hubProtocols.Count > 1 && string.Equals(protocol.Name, "blazorpack", StringComparison.OrdinalIgnoreCase))
```
Comment (anurse) [inline src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs]: Comment as to *why* we're skipping blazor pack :)

File: src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        public DefaultHubMessageSerializer(IHubProtocolResolver hubProtocolResolver, IList<string> globalSupportedProtocols, IList<string> hubSupportedProtocols)
+        {
+            var supportedProtocols = hubSupportedProtocols ?? globalSupportedProtocols ?? Array.Empty<string>();
+            foreach (var protocolName in supportedProtocols)
+            {
+                // blazorpack is meant to only be used by the ComponentHub
+                // We remove it from the list here except when the Hub has a single protocol that is "blazorpack" because
+                // that identifies it as the ComponentHub
+                if (supportedProtocols.Count > 1 && string.Equals(protocolName, "blazorpack", StringComparison.OrdinalIgnoreCase))
```
Comment (pranavkm) [inline src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs]: 😮

Comment (anurse) [reply src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs]: Yeah, it's a bit of a hack :). We want something that allows the Azure SignalR SDK to work without having to wait for any ASP.NET Core release though (and will work with existing 3.0 apps).

File: test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerFacts.cs [hunk 4]
```diff
@@ -34,6 +37,9 @@ public class ServiceLifetimeManagerFacts
                 },
                 NullLogger<DefaultHubProtocolResolver>.Instance);
 
+        private static readonly IOptions<HubOptions> _hubOptions = Options.Create(new HubOptions() { SupportedProtocols = new List<string>() { "json", "messagepack" } });
+        private static readonly IOptions<HubOptions<TestHub>> _hubTOptions = Options.Create(new HubOptions<TestHub>() { SupportedProtocols = new List<string>() { "json", "messagepack" } });
```
Comment (halter73) [inline test/Microsoft.Azure.SignalR.Tests/ServiceLifetimeManagerFacts.cs]: For a second I thought `_hubTOptions` related to strongly typed hubs. `_globalHubOptions` and `_localHubOptions` might make these tests slightly easier to read.

**Comment by vicancy** [comment]
Thanks, @BrennanConroy . The code change looks all good to me. Please target to `dev` branch though. `dev` branch acts as the current working branch and `master` as the release branch. We will merge `dev` changes to `master` to release 1.1.2.

**Comment by BrennanConroy** [comment]
Re-targeted the branch

File: src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs (hunk 3 above)
Comment (davidfowl) [reply src/Microsoft.Azure.SignalR/Internals/DefaultHubMessageSerializer.cs]: 😢

**Pull request merged**